### ISSUE TITLE
Gradient accumulation incompatible with current order of operations

### ISSUE THREAD

**Issue by mcgibbon** [ai2cm/ace#544]
I've stumbled into an issue for implementing gradient accumulation. See [these minimal unit tests](https://github.com/ai2cm/full-model/blob/5e081a5c774d421ff56d40c4374738779c7373e8/fme/fme/test_torch.py) showing this behavior.

`loss.backward()` can only be called multiple times to accumulate gradients if it actually doesn't change the result (tests 1 and 2). In other words, it will not detach gradients between steps. Instead, it raises an exception saying part of the graph the loss depends on has been discarded.

If we explicitly detach the tensors, we can apply gradient accumulation (test 3). This is because the `loss.backward()` calls themselves aren't doing anything magical - we've already explicitly detached the gradients. 

However, if we take these detached steps and concatenate them into a single tensor, all steps now depend on all steps even if we slice that resulting tensor in time later (test 4).

To get around this, I would propose changing the return type of `predict` away from BatchData, and instead returning a type that stores steps as a list of separate tensors. For ease of use, this new type could still expose the same API as BatchData, lazily storing time-concatenated tensors, though this will result in two copies of the data being stored. Or it should be sufficient to time-concatenate the tensors once when it gets passed to the aggregator.

This solution is motivated by our desire to avoid computing the loss in `step` or `predict`.

**Comment by mcgibbon** [comment]
If we must compute the loss in `predict` and/or `step` instead, I'd suggest we do it via a step_callback: Callable that takes in a foo(state: TensorMapping, step: int) so we can still avoid actually putting optimization code in these functions and instead dependency-inject it.

**Comment by jpdunc23** [comment]
I have a strong preference for the `step_callback` approach over creating a new data structure at this stage, especially given the more general usefulness of the `step_callback` feature beyond this particular use case.

**Comment by oliverwm1** [comment]
My first reaction is that the generator approach for the predict method that I proposed would have been a nice way to handle this (i.e. I tend to agree that the time concatenation is not a responsibility that the predict method should have). That said—I also agree we don't want to make a big change to this methods output type right now.

> especially given the more general usefulness of the step_callback feature beyond this particular use case.

What other use cases are you thinking about? Would be good to make sure we consider those to if adding this feature.

**Comment by jpdunc23** [comment]
> What other use cases are you thinking about? Would be good to make sure we consider those to if adding this feature.

Adding a `step_callback` feature has the potential to enable a number of experiments that are currently cumbersome to do or require new bespoke features hard-coded onto `SingleModuleStepper`. Jeremy may have other ideas, but as I envision it we could insert any arbitrary code to run either before (`pre_step_callback`) or after (`post_step_callback`) the existing `step` runs. 

For example, this could make it very easy to do the sort of experiment that Alistair asked for where we remove the salinity-correlated precip bias during inference, e.g. by creating a post-step callback that uses a 2D mask to replace those biased regions with the global precip avg. or some other value. You could do this without modifying an existing corrector and this experiment could just live on a branch / doesn't need to be merged into main unless more generally useful to others as part of a step callback library.

Normalization, tensor packing, corrector, and ocean could all be moved to step callbacks as well. In particular, corrector operations could become more modular so that one can do `force_positive` and other corrections that don't require the atmosphere vertical coord. Slab ocean and prescribed SST could be separated. In coupled training, I could use a custom SST prescriber if I wanted to. Etc.

**Comment by mcgibbon** [comment]
OK, I'll look at resolving this with the step callback route. I think normalization and tensor packing are somewhat intrinsic to taking a step from denormalized TensorMapping and should be more exposed than a callback, but post-module operations like the corrector and ocean could definitely be moved to one. I think we effectively did this in the HEALPix branch just so the callback could be shared between different types of stepping.

**Comment by oliverwm1** [comment]
Actually thinking about it a bit more, you could handle your issue with existing code by just calling the `predict` method multiple times within `train_on_batch`. Slightly awkward, but a possible work around without needing to change implementation of predict.

**Comment by oliverwm1** [comment]
> My first reaction is that the generator approach for the predict method that I proposed would have been a nice way to handle this (i.e. I tend to agree that the time concatenation is not a responsibility that the predict method should have).

Any thoughts on this as a long term strategy? It still feels to me like the right way to handle this—even if we're not going to implement it right now.

**Comment by jpdunc23** [comment]
> Actually thinking about it a bit more, you could handle your issue with existing code by just calling the `predict` method multiple times within `train_on_batch`. Slightly awkward, but a possible work around without needing to change implementation of predict.

True, we could try this approach. We already include a loop over steps in `train_on_batch` to compute the individual step losses, so this wouldn't even add a lot of changes. My main concern is whether we'd be incurring too much extra overhead by calling `predict_paired` multiple times, e.g. the extra calls to `remove_initial_condition` and `get_end` that we would now need for each step.

**Comment by mcgibbon** [comment]
> I tend to agree that the time concatenation is not a responsibility that the predict method should have

While I suggested removing the concatenation as a way around this technical issue, I don't necessarily think predict should not do time concatenation in and of itself.

I could see doing something like this to provide the interface you're suggesting while keeping predict intact, but using it internally. For example, if we refactored `step` to do `predict` but for the smallest unit of prediction (could be multiple steps, for multi-out-step models), then we could wrap `step` in Looper to provide an iterator, and use that iterator internally in `predict` (whose code would then just be iteration and concatenation). This would provide a way to solve the technical issue here by reusing that iterator in `train_on_batch` instead of using `predict` (very similar to "just calling the `predict` method multiple times within `train_on_batch`").

If this ended up being useful for inference we could maybe use it directly and remove predict, but there would be significant performance issues to deal with.

**Comment by oliverwm1** [comment]
> If this ended up being useful for inference we could maybe use it directly and remove predict, but there would be significant performance issues to deal with.

Yeah, to be clear, I still think we may want to keep some N number of steps in memory (and concat them) before passing on to the aggregator/writer. Just that this should be done by the Looper instead of in the predict method itself.
